README.md
# Yohoushi Client [![Build Status](https://secure.travis-ci.org/yohoushi/yohoushi-client.png?branch=master)](http://travis-ci.org/yohoushi/yohoushi-client) [![Dependency Status](https://gemnasium.com/yohoushi/yohoushi-client.png)](https://gemnasium.com/yohoushi/yohoushi-client)

testing ruby: 1.9.3, 2.0.0; Yohoushi: >= 0.62.3

## About Yohoushi Client

yohoushi-client is a ruby client library for Yohoushi API where [yohoushi](http://yohoushi.github.io/yohoushi/) is a visualization graph tool.

## USAGE

    gem install yohoushi-client

See [examples](examples) directory.

## ChangeLog

See [CHANGELOG.md](CHANGELOG.md) for details.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new [Pull Request](../../pull/new/master)

## Copyright

Copyright (c) 2013 DeNA Co., Ltd. See [LICENSE](LICENSE) for details.


CHANGELOG.md
# 0.0.2 (2013/08/29)

Enhancement:

* lstrip / from the given path

# 0.0.1 (2013/08/29)

Initial Release



spec/spec_helper.rb
# encoding: utf-8
require "bundler/setup"

ENV['MOCK'] ||= 'on'
require "pry"
require "pry-nav"
require 'yohoushi-client'
require 'webmock/rspec'
require 'cgi'
WebMock.allow_net_connect! if ENV['MOCK'] == 'off'

ROOT = File.dirname(__FILE__)
Dir[File.expand_path("support/**/*.rb", ROOT)].each {|f| require f }

RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.run_all_when_everything_filtered = true
end


spec/support/mock.rb
# -*- encoding: utf-8 -*-

shared_context "stub_get_graph" do
  proc = Proc.new do
    stub_request(:get, "#{base_uri}/api/graphs/#{escape graph['path']}").
    to_return(:status => 200, :body => graph_example.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_post_graph" do
  include_context "stub_get_graph"
  before do
    stub_request(:post, "#{base_uri}/api/graphs/#{escape graph['path']}").
    to_return(:status => 200, :body => { "error" => 0, "data" => graph_example }.to_json)
  end
end


spec/support/helper.rb
module Yohoushi::SpecHelper
  def escape(str)
    URI.escape(lstrip(str, "/")) if str
  end

  def lstrip(string, substring)
    string.index(substring) == 0 ? string[substring.size..-1] : string
  end

  def base_uri
    'http://localhost:4804'
  end

  def client
    Yohoushi::Client.new(base_uri)
  end

  def graph_example
    {
      "number"=>0,
      "llimit"=>-1000000000,
      "mode"=>"gauge",
      "stype"=>"AREA",
      "adjustval"=>"1",
      "meta"=>"",
      "service_name"=>"multiforecast",
      "gmode"=>"gauge",
      "color"=>"#cc6633",
      "created_at"=>"2013/02/02 00:41:11",
      "section_name"=>"foo%2Fbar%2Ffoo",
      "ulimit"=>1000000000,
      "id"=>1,
      "graph_name"=>"bar",
      "description"=>"",
      "sulimit"=>100000,
      "unit"=>"",
      "sort"=>0,
      "updated_at"=>"2013/02/02 02:32:10",
      "adjust"=>"*",
      "path"=>"foo/bar/foo/bar",
      "type"=>"AREA",
      "sllimit"=>-100000,
      "md5"=>"3c59dc048e8850243be8079a5c74d079"
    }
  end
end


lib/yohoushi/cli.rb
# -*- encoding: utf-8 -*-
require 'thor'
require 'yohoushi-client'

class Yohoushi::CLI < Thor
  class_option :silent, :aliases => ["-S"], :type => :boolean

  desc 'post <json> <url>', 'Post a parameter to a graph'
  long_desc <<-LONGDESC
    Post a parameter to a graph

    ex)
    $ yohoushi-client post '{"number":0}' http://localhost:4804/api/graphs/foo/bar
  LONGDESC
  def post(json, url)
    uri = URI.parse(url)
    client = client(uri)
    path = graph_path(uri.path)
    begin
      res = client.post_graph(path, JSON.parse(json))
      $stdout.puts res unless @options['silent']
    rescue => e
      $stderr.puts "\tclass:#{e.class}\t#{e.message}"
    end
  end

  no_tasks do
    def graph_path(path)
      path.split("/api/graphs/").last
    end

    def client(uri)
      Yohoushi::Client.new("#{uri.scheme}://#{uri.host}:#{uri.port}")
    end
  end
end



lib/yohoushi/client.rb
# -*- encoding: utf-8 -*-
require 'httpclient'
require 'json'
require 'cgi'
require 'pp'

module Yohoushi
  class Error < StandardError; end
  class NotFound < Error; end
  class AlreadyExists < Error; end

  class Client
    attr_accessor :debug
    attr_accessor :client
    attr_reader   :debug_dev
    attr_reader   :base_uri

    # @param [String] base_uri The base uri of Yohoushi
    def initialize(base_uri = 'http://127.0.0.1:4804')
      @base_uri = base_uri
    end

    def client
      @client ||= HTTPClient.new
    end

    # set the `debug_dev` attribute of HTTPClient
    # @param [IO] debug_dev such as STDOUT
    def debug_dev=(debug_dev)
      client.debug_dev = debug_dev
    end

    def last_response
      @res
    end

    # GET the JSON API
    # @param [String] path
    # @return [Hash] response body
    def get_json(path)
      @res = client.get("#{@base_uri}#{path}")
      handle_error(@res)
      JSON.parse(@res.body)
    end

    # POST the JSON API
    # @param [String] path
    # @param [Hash] data 
    # @return [Hash] response body
    def post_json(path, data = {})
      pp data if @debug
      @res = client.post("#{@base_uri}#{path}", data)
      handle_error(@res)
      JSON.parse(@res.body)
    end

    # Post parameters to a graph, POST /api/graphs/:path
    # @param [String] path
    # @param [Hash] params The POST parameters. 
    # @return [Hash] the error code and graph property
    # @example
    # {"error":0,
    # "data":{
    #   "number":"1",
    #   "llimit":"-1000000000",
    #   "mode":"gauge",
    #   "stype":"AREA",
    #   "adjustval":"1",
    #   "meta":"",
    #   "service_name":"multiforecast",
    #   "gmode":"gauge",
    #   "color":"#cc6633",
    #   "created_at":"2013/08/29 11:58:15",
    #   "section_name":"foo%2Fbar%2Ffoo",
    #   "ulimit":"1000000000",
    #   "id":"2",
    #   "graph_name":"bar",
    #   "description":"",
    #   "sulimit":"100000",
    #   "unit":"","sort":"0",
    #   "updated_at":"2013/08/29 11:58:15",
    #   "adjust":"*",
    #   "path":"foo/bar/foo/bar",
    #   "type":"AREA",
    #   "sllimit":"-100000",
    #   "md5":"c81e728d9d4c2f636f067f89cc14862c"}
    def post_graph(path, params)
      post_json("/api/graphs/#{escape path}", params)
    end

    # Get parameters from a graph, GET /api/graphs/:path
    # @param [String] path
    # @return [Hash] the graph property
    # @example
    # {
    #   "number":"1",
    #   "llimit":"-1000000000",
    #   "mode":"gauge",
    #   "stype":"AREA",
    #   "adjustval":"1",
    #   "meta":"",
    #   "service_name":"multiforecast",
    #   "gmode":"gauge",
    #   "color":"#cc6633",
    #   "created_at":"2013/08/29 11:58:15",
    #   "section_name":"foo%2Fbar%2Ffoo",
    #   "ulimit":"1000000000",
    #   "id":"2",
    #   "graph_name":"bar",
    #   "description":"",
    #   "sulimit":"100000",
    #   "unit":"","sort":"0",
    #   "updated_at":"2013/08/29 11:58:15",
    #   "adjust":"*",
    #   "path":"foo/bar/foo/bar",
    #   "type":"AREA",
    #   "sllimit":"-100000",
    #   "md5":"c81e728d9d4c2f636f067f89cc14862c"}
    def get_graph(path)
      get_json("/api/graphs/#{escape path}")
    end

    private

    def escape(str)
      URI.escape(lstrip(str, "/")) if str
    end

    def lstrip(string, substring)
      string.index(substring) == 0 ? string[substring.size..-1] : string
    end

    def handle_error(res)
      case res.status
      when 200
      when 201
      when 404
        raise NotFound.new(error_message(res))
      when 409
        raise AlreadyExists.new(error_message(res))
      else
        raise Error.new(error_message(res))
      end
    end

    def error_message(res)
      "status:#{res.status}\turi:#{res.http_header.request_uri.to_s}\tmessage:#{res.body}"
    end
  end
end

